core/finality-grandpa/src/lib.rs

```diff
@@ -216,6 +215,30 @@ impl From<ClientError> for Error {
 	}
 }
 
+/// A stream used by NetworkBridge in its implementation of Network.
+pub struct NetworkStream {
+	inner: Option<mpsc::UnboundedReceiver<Vec<u8>>>,
+	outer: oneshot::Receiver<mpsc::UnboundedReceiver<Vec<u8>>>
+}
+
+impl Stream for NetworkStream {
+	type Item = Vec<u8>;
+	type Error = ();
+
+	fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+		if let Some(ref mut inner) = self.inner {
+			return inner
+				.poll()
+		}
+		if let Ok(futures::Async::Ready(mut inner)) = self.outer.poll() {
```

Comment:
If the `oneshot::Receiver` is returning `Error(Canceled)`, this will stuck forever.

Reply:
How could we handle the error? An error would indicate network has shut down, and the entire system is likely shutting down. We could panic on an `Err` I guess...

Reply:
I don't know what the error indicates. Maybe we should just throw `Err(())`.

Reply:
We can propagate an appropriate error.